cmd/version.go

```diff
@@ -16,11 +16,64 @@ package cmd
 
 //go:generate ./mkversion.sh
 
-// Version will be overwritten at build-time via mkversion.sh
-var Version = "unknown"
+import (
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+)
 
-func MockVersion(version string) (restore func()) {
-	old := Version
-	Version = version
-	return func() { Version = old }
+var (
+	// Version will be overwritten at build-time via mkversion.sh
+	Version = "unknown"
+
+	pid2ProcPath = "/proc/2/status"
+	selfPid      = os.Getpid()
+
+	containerOnce    sync.Once
+	containerRuntime bool = true
+)
+
+// Containerised returns true if we are running inside a container runtime
+// such as lxd or Docker. The detection is only intended for Linux host
+// environments as it looks for the Linux kernel kthreadd process, started
+// by the Linux kernel after init. Kernel initialised processes do not have
+// a parent as its not a child of init (PID1) and as a result has its ppid
+// set to zero. We use this property to detect the absence of a container
+// runtime. If /proc is not mounted, the detection will assume its a
+// container, so make sure early mounts are completed before running this
+// check on an unconfined host.
+func Containerised() bool {
```

Comment:
Let's refactor this function to breakout the detection logic from the synchronisation logic so that it can be tested properly.

```
func Containerised() bool {
	containerOnce.Do(func() {
		containerRuntime, _ = detectContainerised(pid2ProcPath)
	})
	return containerRuntime
}

func detectContainerised(procPath string) (bool, error) {
	s, err := ioutil.ReadFile(procPath);
	if err != nil {
		return true, err
	}
	lines := strings.Split(string(s), "\n")
	for _, l := range lines {
		kv := strings.Split(l, "\t")
		if zeroPPid(kv) {
			return false, nil
		}
	}
	return true, nil
}
```

Reply:
Split it out, much better, thanks.